# Pandas DataFrames 101 - Learning Notebook

----

<img src="https://miro.medium.com/v2/resize:fit:1400/1*Q55X1gfHjKi1knx96UVdZQ.png" alt="Description" width="450">

# What you will learn in this notebook

In this notebook we will cover the following:

1. Dataframes 
    1. Reading data into a dataframe
    2. Common methods for a quick overview
    3. Getting the index and column values
2. Previewing and describing a DataFrame
    1. Previewing the DataFrame or part of it
    2. Retrieving DataFrame information
3. Reading data from files into pandas dataframes
4. Writing data from pandas into files

# Imports

In [1]:
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"

In [2]:
import pandas as pd
import os

# What is a DataFrame

A 2D, potentially heterogenous, tabular structure. The documentation on DataFrame is available on the `pandas.DataFrame` [documentation page](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.html) 

Let's start by reading some dataset on heart disease ([source](https://archive.ics.uci.edu/dataset/45/heart+disease)). 

In [3]:
df = pd.read_csv('https://raw.githubusercontent.com/Hospital-Da-Luz-Learning-Health/MLCatolica24/main/Aula%203%20-%20Data%20Analysis%20%26%20Statistics/Pandas%20DataFrames%20101/data/heart_disease_cleaned.csv')
df = df.set_index('id')

Let's take a sneak peak at this data:

In [4]:
df.head(3)
#df.tail(3)
df.shape

age   sex                cp  trestbps  chol  fbs  \
id                                                      
192   43  male      asymptomatic       132   247    1   
189   69  male  non-anginal pain       140   254    0   
0     63  male    typical angina       145   233    1   

                          restecg  thalach  exang  oldpeak        slope   ca  \
id                                                                             
192  left ventricular hypertrophy      143      1      0.1          fat  NaN   
189  left ventricular hypertrophy      146      0      2.0          fat  3.0   
0    left ventricular hypertrophy      150      0      2.3  downsloping  0.0   

                  thal Diagnosed  
id                                
192  reversible defect      >50%  
189  reversible defect      >50%  
0         fixed defect      <50%

(303, 14)

Here's what we know about this dataset:

- contains 303 rows (observations)
- 14 columns

And some documentation we found about each column:

1. **Age:** age in years
2. **Sex:** sex (1 = male; 0 = female)
3. **Cp:** chest pain type
   - Value 1: typical angina
   - Value 2: atypical angina
   - Value 3: non-anginal pain
   - Value 4: asymptomatic
4. **Trestbps:** resting blood pressure (in mm Hg on admission to the hospital)
5. **Chol:** serum cholesterol in mg/dl
6. **Fbs:** fasting blood sugar > 120 mg/dl (1 = true; 0 = false)
7. **Restecg:** resting electrocardiographic results
   - Value 0: normal
   - Value 1: having ST-T wave abnormality (T wave inversions and/or ST elevation or depression of > 0.05 mV)
   - Value 2: showing probable or definite left ventricular hypertrophy by Estes' criteria
8. **Thalach:** maximum heart rate achieved
9. **Exang:** exercise induced angina (1 = yes; 0 = no)
10. **Oldpeak:** ST depression induced by exercise relative to rest
11. **Slope:** the slope of the peak exercise ST segment
    - Value 1: upsloping
    - Value 2: flat
    - Value 3: downsloping
12. **Ca:** number of major vessels (0-3) colored by flourosopy
13. **Thal:** 
    - 3 = normal
    - 6 = fixed defect
    - 7 = reversible defect
14. **Diagnosed:** diagnosis of heart disease (angiographic disease status)
    - Value 0: < 50% diameter narrowing
    - Value 1: > 50% diameter narrowing


# Selecting columns and index

In [5]:
df['age'] # selects a column and return a pandas series

id
192    43
189    69
0      63
23     58
284    61
       ..
9      53
109    39
185    63
111    56
36     43
Name: age, Length: 303, dtype: int64

In [6]:
df.loc[:,'age'] # does the same thing. "select all rows, and only the column <age>"

id
192    43
189    69
0      63
23     58
284    61
       ..
9      53
109    39
185    63
111    56
36     43
Name: age, Length: 303, dtype: int64

We can access the index like so:

In [7]:
df.index

Index([192, 189,   0,  23, 284, 177,  89, 290, 127,  56,
       ...
       163, 169, 186, 244, 139,   9, 109, 185, 111,  36],
      dtype='int64', name='id', length=303)

And this works like a list, so let's see who is the patient in the 41st line

In [8]:
df.index[40]

235

> Give me the `age` of the patient whose `id is 235`

In [9]:
df.loc[235,'age']

54

> Give me the `slope` and `cp` of patient `177`

In [10]:
df.loc[177,['slope','cp']]

slope             fat
cp       asymptomatic
Name: 177, dtype: object

> Give me the `sex` and `thal` of patients `23,199, and 102`

In [11]:
df.loc[[23,199,102],['sex','thal']]

sex               thal
id                            
23     male  reversible defect
199    male             normal
102  female             normal

# Descriptive Statistics

What type of data do we have?

In [12]:
df.dtypes

age            int64
sex           object
cp            object
trestbps       int64
chol           int64
fbs            int64
restecg       object
thalach        int64
exang          int64
oldpeak      float64
slope         object
ca           float64
thal          object
Diagnosed     object
dtype: object

## Analyzing numerical data

In [13]:
df['age'].head(3)

id
192    43
189    69
0      63
Name: age, dtype: int64

In [14]:
df['age'].mean()

54.43894389438944

In [15]:
df['age'].median()

56.0

In [16]:
df['age'].max()

77

In [17]:
df['age'].min()

29

[Here are all descriptive statistics you can easily compute from a Pandas series](https://pandas.pydata.org/pandas-docs/version/0.20.2/api.html#api-dataframe-stats)

You can also use `.describe()` to get the most common stats of a numerical column

In [18]:
df['age'].describe()

count    303.000000
mean      54.438944
std        9.038662
min       29.000000
25%       48.000000
50%       56.000000
75%       61.000000
max       77.000000
Name: age, dtype: float64

in case you want a better display, you can convert a pandas series to dataframe with `.to_frame()`

In [19]:
# in case you want a better display, you can convert a pandas series to dataframe 
df['age'].describe().to_frame()

age
count  303.000000
mean    54.438944
std      9.038662
min     29.000000
25%     48.000000
50%     56.000000
75%     61.000000
max     77.000000

> Go to exercises notebook and solve exercises 1 through 4

## Analyzing categorical data

In [20]:
df.dtypes

age            int64
sex           object
cp            object
trestbps       int64
chol           int64
fbs            int64
restecg       object
thalach        int64
exang          int64
oldpeak      float64
slope         object
ca           float64
thal          object
Diagnosed     object
dtype: object

> What is the distribution of `x`?

In [21]:
df['sex'].value_counts()

sex
male      206
female     97
Name: count, dtype: int64

In [22]:
df['sex'].value_counts(normalize=True)

sex
male      0.679868
female    0.320132
Name: proportion, dtype: float64

Can we round these numbers? (let's see with chatgpt)

----

I would like to look at the distribution of all categorical columns

In [23]:
df.select_dtypes('object')

sex                cp                       restecg        slope  \
id                                                                         
192    male      asymptomatic  left ventricular hypertrophy          fat   
189    male  non-anginal pain  left ventricular hypertrophy          fat   
0      male    typical angina  left ventricular hypertrophy  downsloping   
23     male  non-anginal pain  left ventricular hypertrophy    upsloping   
284    male      asymptomatic                        normal    upsloping   
..      ...               ...                           ...          ...   
9      male      asymptomatic  left ventricular hypertrophy  downsloping   
109    male      asymptomatic                        normal          fat   
185  female   atypical angina                        normal    upsloping   
111    male      asymptomatic  left ventricular hypertrophy          fat   
36     male      asymptomatic  left ventricular hypertrophy          fat   

                  thal Diagnosed  
id                                
192  reversible defect      >50%  
189  reversible defect      >50%  
0         fixed defect      <50%  
23   reversible defect      >50%  
284  reversible defect      >50%  
..                 ...       ...  
9    reversible defect      >50%  
109  reversible defect      >50%  
185             normal      <50%  
111             normal      >50%  
36   reversible defect      >50%  

[303 rows x 6 columns]

In [24]:
cat_columns = df.select_dtypes('object').columns

In [25]:
for col in cat_columns:
    df[col].value_counts(normalize=True).round(2).to_frame()

proportion
sex               
male          0.68
female        0.32

proportion
cp                          
asymptomatic            0.48
non-anginal pain        0.28
atypical angina         0.17
typical angina          0.08

proportion
restecg                                 
normal                              0.50
left ventricular hypertrophy        0.49
ST-T wave abnormality               0.01

proportion
slope                  
upsloping          0.47
fat                0.46
downsloping        0.07

proportion
thal                         
normal                   0.55
reversible defect        0.39
fixed defect             0.06

proportion
Diagnosed            
<50%             0.54
>50%             0.46

# Masks

<img src="https://images7.memedroid.com/images/UPLOADED669/640ef2a4662e1.jpeg" alt="Description" width="450">

Masks are great because they allow us to select only part of data you want

In [26]:
mask = df['age'] > 55

In [27]:
mask

id
192    False
189     True
0       True
23      True
284     True
       ...  
9      False
109    False
185     True
111     True
36     False
Name: age, Length: 303, dtype: bool

In [28]:
mask.value_counts()

age
True     152
False    151
Name: count, dtype: int64

and now we use this mask to select only the rows with `True`

In [29]:
df.loc[mask,:]

age     sex                cp  trestbps  chol  fbs  \
id                                                        
189   69    male  non-anginal pain       140   254    0   
0     63    male    typical angina       145   233    1   
23    58    male  non-anginal pain       132   224    0   
284   61    male      asymptomatic       148   203    0   
177   56    male      asymptomatic       132   184    0   
..   ...     ...               ...       ...   ...  ...   
251   58    male      asymptomatic       146   218    0   
163   58  female      asymptomatic       100   248    0   
244   60  female  non-anginal pain       120   178    1   
185   63  female   atypical angina       140   195    0   
111   56    male      asymptomatic       125   249    1   

                          restecg  thalach  exang  oldpeak        slope   ca  \
id                                                                             
189  left ventricular hypertrophy      146      0      2.0          fat  3.0   
0    left ventricular hypertrophy      150      0      2.3  downsloping  0.0   
23   left ventricular hypertrophy      173      0      3.2    upsloping  2.0   
284                        normal      161      0      0.0    upsloping  1.0   
177  left ventricular hypertrophy      105      1      2.1          fat  1.0   
..                            ...      ...    ...      ...          ...  ...   
251                        normal      105      0      2.0          fat  1.0   
163  left ventricular hypertrophy      122      0      1.0          fat  0.0   
244                        normal       96      0      0.0    upsloping  0.0   
185                        normal      179      0      0.0    upsloping  2.0   
111  left ventricular hypertrophy      144      1      1.2          fat  1.0   

                  thal Diagnosed  
id                                
189  reversible defect      >50%  
0         fixed defect      <50%  
23   reversible defect      >50%  
284  reversible defect      >50%  
177       fixed defect      >50%  
..                 ...       ...  
251  reversible defect      >50%  
163             normal      <50%  
244             normal      <50%  
185             normal      <50%  
111             normal      >50%  

[152 rows x 14 columns]

In [30]:
# let's confirm the above
df.loc[mask,'age'].min()

56

One cool thing about this is that you can better analyze your data, check the following example:

In [31]:
df.loc[mask,'Diagnosed'].value_counts(normalize=True)

Diagnosed
>50%    0.592105
<50%    0.407895
Name: proportion, dtype: float64

Does this pattern maintain for people younger than 55?

In [32]:
df.loc[~mask,:].head(3)

age     sex                cp  trestbps  chol  fbs  \
id                                                        
192   43    male      asymptomatic       132   247    1   
89    51  female  non-anginal pain       130   256    0   
127   54    male      asymptomatic       110   239    0   

                          restecg  thalach  exang  oldpeak      slope   ca  \
id                                                                           
192  left ventricular hypertrophy      143      1      0.1        fat  NaN   
89   left ventricular hypertrophy      149      0      0.5  upsloping  0.0   
127                        normal      126      1      2.8        fat  1.0   

                  thal Diagnosed  
id                                
192  reversible defect      >50%  
89              normal      <50%  
127  reversible defect      >50%

In [33]:
df.loc[~mask,'Diagnosed'].value_counts(normalize=True)

Diagnosed
<50%    0.675497
>50%    0.324503
Name: proportion, dtype: float64

Very different!

We can also join together masks as well with logical operations

In [34]:
mask_age_above_55 = df['age'] > 55
mask_cp_asymptomatic = df['cp'] == 'asymptomatic'

In [35]:
df.loc[mask_age_above_55 & mask_cp_asymptomatic,:].head(3)

age   sex            cp  trestbps  chol  fbs  \
id                                                  
284   61  male  asymptomatic       148   203    0   
177   56  male  asymptomatic       132   184    0   
96    59  male  asymptomatic       110   239    0   

                          restecg  thalach  exang  oldpeak      slope   ca  \
id                                                                           
284                        normal      161      0      0.0  upsloping  1.0   
177  left ventricular hypertrophy      105      1      2.1        fat  1.0   
96   left ventricular hypertrophy      142      1      1.2        fat  1.0   

                  thal Diagnosed  
id                                
284  reversible defect      >50%  
177       fixed defect      >50%  
96   reversible defect      >50%

# GroupBy

`Groupby` is useful when you want to ask a question about the dataset, and you want the answer stratified by groups!

<p align="center">
  <img src="media/gorupby.png" alt="image credit"><br>
  <em>Image adapted from <a href="https://towardsdatascience.com/11-examples-to-master-pandas-groupby-function-86e0de574f38">https://towardsdatascience.com/11-examples-to-master-pandas-groupby-function-86e0de574f38</a></em>
</p>

Let's take a look at the following example:

> "I want the average age of my patients, but stratified by gender (i.e. for males and females separately)"

We've learned with `masks` that we can do stuff like this already

In [36]:
mask_males = df['sex'] == 'male'

In [37]:
# just checking whether the masks are correct
df.loc[mask_males,:].head(5)

age   sex                cp  trestbps  chol  fbs  \
id                                                      
192   43  male      asymptomatic       132   247    1   
189   69  male  non-anginal pain       140   254    0   
0     63  male    typical angina       145   233    1   
23    58  male  non-anginal pain       132   224    0   
284   61  male      asymptomatic       148   203    0   

                          restecg  thalach  exang  oldpeak        slope   ca  \
id                                                                             
192  left ventricular hypertrophy      143      1      0.1          fat  NaN   
189  left ventricular hypertrophy      146      0      2.0          fat  3.0   
0    left ventricular hypertrophy      150      0      2.3  downsloping  0.0   
23   left ventricular hypertrophy      173      0      3.2    upsloping  2.0   
284                        normal      161      0      0.0    upsloping  1.0   

                  thal Diagnosed  
id                                
192  reversible defect      >50%  
189  reversible defect      >50%  
0         fixed defect      <50%  
23   reversible defect      >50%  
284  reversible defect      >50%

now we compute the `mean` of the age

In [38]:
average_age_males = df.loc[mask_males,'age'].mean()
print('average age of male patients is:', average_age_males)

average age of male patients is: 53.83495145631068


and we repeat for the females

In [44]:
mask_females = df['sex'] == 'female'
average_age_females = df.loc[mask_females,'age'].mean()
print('average_age of female patients is:', average_age_females)

average_age of female patients is: 55.72164948453608


Nice! However programmers like to do more with fewer lines of code.


Using `groupby` instead, we get:

In [40]:
df.groupby('sex')['age'].mean()

sex
female    55.721649
male      53.834951
Name: age, dtype: float64

![](https://i.imgflip.com/8v206r.jpg)

In [41]:
# we can "prettify" the previous results
df.groupby('sex')['age'].mean().to_frame().round(1)

age
sex         
female  55.7
male    53.8

## Let's start to answer a series of questions using `groupby`

In [58]:
df.head(2)

age   sex                cp  trestbps  chol  fbs  \
id                                                      
192   43  male      asymptomatic       132   247    1   
189   69  male  non-anginal pain       140   254    0   

                          restecg  thalach  exang  oldpeak slope   ca  \
id                                                                      
192  left ventricular hypertrophy      143      1      0.1   fat  NaN   
189  left ventricular hypertrophy      146      0      2.0   fat  3.0   

                  thal Diagnosed  
id                                
192  reversible defect      >50%  
189  reversible defect      >50%

### What is the frequency of patients of each cp category

In [54]:
(df.groupby('cp').size()
 .to_frame('Number of Patients')
)

Number of Patients
cp                                  
asymptomatic                     144
atypical angina                   50
non-anginal pain                  86
typical angina                    23

Notice how we didn't have to select a column here because you don't need a specific column to compute the size of a group 

### What is the distribution of `oldpeak` (numerical) by `sex` (category)?

This is how we've already learned to quickly check some main distribution statistics of a numerical column:

In [56]:
df['oldpeak'].describe() # analyzing the column directly

count    303.000000
mean       1.039604
std        1.161075
min        0.000000
25%        0.000000
50%        0.800000
75%        1.600000
max        6.200000
Name: oldpeak, dtype: float64

with `groupby` it's similar:

In [57]:
df.groupby('sex')['oldpeak'].describe()

count      mean       std  min  25%  50%  75%  max
sex                                                       
female   97.0  0.867010  1.117542  0.0  0.0  0.6  1.4  6.2
male    206.0  1.120874  1.174901  0.0  0.0  0.8  1.8  5.6

### What is the distribution of `slope` (category) by `cp` (category)?

In [76]:
df.groupby('cp')['slope'].value_counts()

cp                slope      
asymptomatic      fat            84
                  upsloping      49
                  downsloping    11
atypical angina   upsloping      36
                  fat            12
                  downsloping     2
non-anginal pain  upsloping      48
                  fat            33
                  downsloping     5
typical angina    fat            11
                  upsloping       9
                  downsloping     3
Name: count, dtype: int64

While it seems we have 3 columns in the result above, this is actually just a pandas series whose index has 2 levels.

In [77]:
df.groupby('cp')['slope'].value_counts().index

MultiIndex([(    'asymptomatic',         'fat'),
            (    'asymptomatic',   'upsloping'),
            (    'asymptomatic', 'downsloping'),
            ( 'atypical angina',   'upsloping'),
            ( 'atypical angina',         'fat'),
            ( 'atypical angina', 'downsloping'),
            ('non-anginal pain',   'upsloping'),
            ('non-anginal pain',         'fat'),
            ('non-anginal pain', 'downsloping'),
            (  'typical angina',         'fat'),
            (  'typical angina',   'upsloping'),
            (  'typical angina', 'downsloping')],
           names=['cp', 'slope'])

either way, the results are hard to understand at a glance, so let's "prettify" this

In [82]:
df.groupby('cp')['slope'].value_counts().to_frame('Count') # simply converted the pandas series to a dataframe

Count
cp               slope             
asymptomatic     fat             84
                 upsloping       49
                 downsloping     11
atypical angina  upsloping       36
                 fat             12
                 downsloping      2
non-anginal pain upsloping       48
                 fat             33
                 downsloping      5
typical angina   fat             11
                 upsloping        9
                 downsloping      3

As a dataframe it looks better already but still not good enough

In [83]:
df.groupby('cp')['slope'].value_counts().to_frame('Count').unstack(level='cp')

Count                                                
cp          asymptomatic atypical angina non-anginal pain typical angina
slope                                                                   
downsloping           11               2                5              3
fat                   84              12               33             11
upsloping             49              36               48              9

`unstack()` pivots one of the index levels and has made the table a lot easier to interpret. You can see the [documentation](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.unstack.html) for a more clear explanation and examples

In [78]:
(df
 .groupby('cp')['slope'].value_counts() # what we had so far
 .to_frame('Frequency')
 .unstack(level='cp')
)

Frequency                                                
cp          asymptomatic atypical angina non-anginal pain typical angina
slope                                                                   
downsloping           11               2                5              3
fat                   84              12               33             11
upsloping             49              36               48              9